### PR TITLE
osd: bluestore: kick out onodes from onode_map as soon as they are removed

### PR DESCRIPTION
**Pull request by xiexingguo**
Signed-off-by: xie xingguo <xie.xingguo@zte.com.cn>

### CODE REVIEW

**Comment by jjhuo** [comment]
Please see
https://github.com/liewegas/ceph/commit/bf1f02c1b0b6a5ee6cd651dc80280b80d3044af9

**Comment by xiexingguo** [comment]
@jjhuo  Thank you for your attention. Let's wait and see what is sage's suggestion:-)

File: src/os/bluestore/BlueStore.cc [hunk 1]
```diff
@@ -498,6 +498,17 @@ void BlueStore::OnodeHashLRU::add(const ghobject_t& oid, OnodeRef o)
...
 
+void BlueStore::OnodeHashLRU::remove(OnodeRef o)
+{
+  std::lock_guard<std::mutex> l(lock);
+  dout(30) << __func__ << " " << o->oid << " " << o << dendl;
+  lru_list_t::iterator p = lru.iterator_to(*o);
+  lru.erase(p);
+  ghobject_t oid = o->oid;
+  assert(onode_map.count(oid));
+  onode_map.erase(oid);
```
Comment (liewegas) [inline src/os/bluestore/BlueStore.cc]: no need for the ghobject_t oid local copy

**Comment by liewegas** [comment]
We often have transactions taht remove and then recreate, and we don't want those to be a miss, but i think it's fine because of the ovec in _txc_add_transaction.  Did you run this through ceph_test_objectstore?

**Comment by jjhuo** [comment]
@liewegas kind of interested to know, why did you delete the remove function in the first place?
https://github.com/ceph/ceph/commit/bf1f02c1b0b6a5ee6cd651dc80280b80d3044af9
Were you trying to avoid frequent unordered_map rehashing?

Also, I feel orignal remove function is better implemented.

**Comment by xiexingguo** [comment]
Fixed and pass my local test(script: ceph_test_objectstore), below is the result:
```
==> rm -r store_test_temp_dir
[       OK ] ObjectStore/StoreTest.SetAllocHint/3 (592 ms)
[----------] 160 tests from ObjectStore/StoreTest (3027755 ms total)

[----------] Global test environment tear-down
[==========] 160 tests from 1 test case ran. (3027756 ms total)
[  PASSED  ] 160 tests.
[root@localhost src]# 
```

**Comment by liewegas** [comment]
Oh, I see.  Yeah, I'm inclined to keep the negative entries in the map and let them expire on a normal schedule.

**Comment by xiexingguo** [comment]
OK :-), I am glad that you are still here.

**Pull request closed without merging**